primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -0,0 +1,262 @@
+
+/// Namespaced Merkle Tree where leaves are sorted by a namespace identifier.
+pub trait NamespacedMerkleTreeScheme: AppendableMerkleTreeScheme
+where
+    Self::Element: Namespaced<Self::NamespaceId>,
+{
+    /// Namespace proof type
+    type NamespaceProof: Clone;
+    /// Namespace type
+    type NamespaceId: Namespace;
+
+    /// Returns the entire set of leaves corresponding to a given namespace and
+    /// a completeness proof
+    fn get_namespace_leaves_and_proof(
+        &self,
+        namespace: Self::NamespaceId,
+    ) -> (Vec<Self::Element>, Self::NamespaceProof);
+
+    /// Verifies the completeness proof for a given set of leaves and a
+    /// namespace
+    fn verify_namespace_proof(
+        leaves: &[Self::Element],
+        proof: Self::NamespaceProof,
+        namespace: Self::NamespaceId,
+    ) -> Result<(), PrimitivesError>;
+}
+
+/// NamespacedHasher wraps a standard hash function (implementer of
+/// DigestAlgorithm), turning it into a hash function that tags internal nodes
+/// with namespace ranges
+pub struct NamespacedHasher<H, E, I, T, N>
+where
+    H: DigestAlgorithm<E, I, T>,
+    E: Element + Namespaced<N>,
+    N: Namespace,
+    I: Index,
+    T: NodeValue,
+{
+    phantom1: PhantomData<H>,
+    phantom2: PhantomData<E>,
+    phantom3: PhantomData<I>,
+    phantom4: PhantomData<T>,
+    phantom5: PhantomData<N>,
+}
+
+/// Trait indicating that a leaf has a namespace
+pub trait Namespaced<N: Namespace> {
+    /// Returns the namespace of the leaf
+    fn get_namespace(&self) -> N;
+}
+
+/// Trait indicating that a digest algorithm can bind namespaces
+pub trait BindNamespace<E, I, T, N>: DigestAlgorithm<E, I, T>
+where
+    E: Element,
+    N: Namespace,
+    T: NodeValue,
+    I: Index,
+{
+    /// Generate a commitment that binds a node to a namespace range
+    fn generate_namespaced_commitment(namespaced_hash: NamespacedHash<T, N>) -> T;
+}
+
+/// Trait indiciating that a struct can act as an orderable namespace
+pub trait Namespace:
+    Debug + Clone + CanonicalDeserialize + CanonicalSerialize + Default + Copy + Hash + Ord
+{
+    /// Returns the minimum possible namespace
+    fn min() -> Self;
+    /// Returns the maximum possible namespace
+    fn max() -> Self;
+}
+
+impl Namespace for u64 {
+    fn min() -> u64 {
+        u64::MIN
+    }
+    fn max() -> u64 {
+        u64::MAX
+    }
+}
+
+#[derive(
+    CanonicalSerialize,
+    CanonicalDeserialize,
+    Hash,
+    Copy,
+    Clone,
+    Debug,
+    Ord,
+    Eq,
+    PartialEq,
+    PartialOrd,
+)]
+/// Represents a namespaced internal tree node
+pub struct NamespacedHash<T, N>
+where
+    N: Namespace,
+    T: NodeValue,
+{
+    min_namespace: N,
+    max_namespace: N,
+    hash: T,
+}
+
+impl<T, N> Default for NamespacedHash<T, N>
+where
+    N: Namespace,
+    T: NodeValue,
+{
+    fn default() -> Self {
+        Self {
+            hash: T::default(),
+            max_namespace: <N as Namespace>::min(),
```

Comment:
Min/max is swapped like this so that hashing a namespace hash `NamespaceHash::default()` preserves the namespace range

Reply:
I am actually not sure about this - removed this for now but will revisit it when we start implementing proofs